src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/JsonHubProtocol.java

```diff
@@ -181,7 +206,7 @@ public TransferFormat getTransferFormat() {
 
     @Override
     public String writeMessage(HubMessage hubMessage) {
-        return gson.toJson(hubMessage) + RECORD_SEPARATOR;
+        return internalGgson.toJson(hubMessage) + RECORD_SEPARATOR;
```

Comment:
This is where the problem you reference in the description lies. In .NET we never use the "serialize/deserialize" APIs to write out the protocol messages. We use the writer/reader APIs directly, then use the user-provided serializer to convert the individual payload items (arguments, return values) to JSON and write them out to the final message in the correct location.

I think to do this right we need to convert this to use `JsonWriter` (much like the parser uses `JsonReader`).

Pseudo code something like this:

```
JsonWriter writer = ...;
writer.beginObject();
if (message is a CompletionMessage completionMessage) {
	writer.name("type");
	writer.value(/* Type value for CompletionMessage */)
	// other properties ...

	// Now to serialize the "result"
	writer.name("result");

	// Use the "userGson" serializer to write the object itself
	userGson.toJson(completionMessage.getResult(), completionMessage.getResult().getClass(), writer);
}
// So on and so forth for other message types.
writer.endObject();
```

Reply:
Thanks for your feedback. The rest seems straightforward but I'd like to make sure I understand this one.

So are you suggesting this would better be solved by adding a method like:
`abstract void write(JsonWriter writer);`
to HubMessage.java? And eliminating usages of `internalGson` (as I've named it) entirely?

Reply:
@anurse feel free to correct me if I'm wrong here, but I'm pretty sure the idea is not to change HubMessage.java at all. The idea is to simply change this JsonHubProtocol.writeMessage method to serialize using a JsonWriter and put all the type-specific logic for serializing each HubMessage type to JSON inside of JsonHubProtocol.java itself.

So instead of having an `abstract void write(JsonWriter writer);` implementation on InvocationMessage and calling `message.write(writer)` from JsonHubProtocol.writeMessage, you would have a `if (message instanceof InvocationMessage) { ... } else if (message instanceof ...`. It's less object-oriented, but it will make adding non-JSON protocols like MessagePack easier in the future. Instead of every HubMessage needing to know how to serialize using JsonWriter, MessagePackWriter and whatever else, all the serialization logic stays within the protocol class this way.

Reply:
As a layman, I'd like to mention that if binary protocols were to be introduced to the Java client then protocol buffers would be in consideration for several reasons, and (assuming you agree) that would need a further layer of abstraction.

Reply:
Not quite, though it sounds like you're close. I don't think a `HubMessage` should know how to serialize itself to JSON. `HubMessage` is just a data object carrying information. The JsonHubProtocol is the one that knows how to serialize messages. It's fine for it to know about each individual subclass of `HubMessage` since it has to know that anyway in order to properly serialize/deserialize.

I'm basically suggesting you put the implementations of all those `write(JsonWriter)` methods in `JsonHubProtocol` itself. You would then check the type of the incoming message, find the appropriate "writer" method and call it (casting the message). Each of those methods would **manually** write out each property (using the `JsonWriter` APIs), with one important exception: When it comes to writing out `CompletionMessage.getResult()`, the items in `InvocationMessage.getArguments()` or `StreamInvocationMessage.getArguments()` and `StreamItem.getItem()`, we use the user-provided `Gson` to do serialize the object directly into the `JsonWriter`.

Reply:
@anurse I think I understand, but it sounds like there should be a mapping between message types and protocol writers? So:

`[{binary,message},{json,message},{binary,streamItem},{json,streamItem}]`

And you choose a mapping between those types? Trouble is there's only one protocol in the current implementation so I'm not sure how to fit in :)

TBH I feel a bit out of my depth at this point because I'm not sure of your requirements. I can swing a VC call sometime tomorrow if you like?

Reply:
> that if binary protocols were to be introduced to the Java client then protocol buffers would be in consideration for several reasons, and (assuming you agree) that would need a further layer of abstraction.

That's precisely why it's the `HubProtocol` that should be responsible for serialization, not the `HubMessage`. The `MessagePackHubProtocol` (when/if it is implemented) will have a similar tree of `if`s and will do different serialization logic.

I don't think **new** abstraction is needed here. It's entirely reasonable to couple `HubProtocol`s **to** `HubMessage`s. What we don't want to do is couple `HubMessage`s **to** `HubProtocol`s.